torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -707,22 +707,30 @@ Tensor prod_backward(
   if (input.dim() == 0) {
     return grad;
   }
+  if (input.numel() == 0) {
+    return input;
+  }
   dim = at::maybe_wrap_dim(dim, input.sizes().size());
   if (!keepdim && input.dim() != 1) {
     grad = grad.unsqueeze(dim);
     result = result.unsqueeze(dim);
   }
   if (input.is_meta()) {
-    return prod_safe_zeros_backward(grad, input, dim);
+    return prod_safe_zeros_backward(grad, input, dim).view_as(input);
   }
 
   Tensor zero_mask = (input == 0);
   Tensor slice_zero_count = zero_mask.sum(dim, true);
-  int64_t total_zeros = slice_zero_count.sum().item<int64_t>();
-  if (total_zeros == 0) {
-    return grad * (result / input).conj();
+  auto total_zeros = slice_zero_count.sum();
+  if (isTensorSubclassLike(total_zeros)) {
+    // For Composite Compliance, always take the safer (and slower) path
+    return prod_safe_zeros_backward(grad, input, dim).view_as(input);
   } else {
-    return prod_safe_zeros_backward(grad, input, dim);
+    if (total_zeros.item<int64_t>() == 0) {
+      return grad * (result / input).conj();
+    } else {
+      return prod_safe_zeros_backward(grad, input, dim).view_as(input);
```

Comment:
Is there an explanation for why the view_as is necessary? If it is always necessary, then would it be better to put the view_as inside the prod_safe_zeros_backward function?

Reply:
Have updated the fix. So we don't need view_as anymore. 
PTAL :)